mlir/include/mlir/Dialect/SCF/Transforms/Patterns.h

```diff
@@ -85,6 +86,10 @@ void populateSCFForLoopCanonicalizationPatterns(RewritePatternSet &patterns);
 ///  * `after` block containing arith.addi
 void populateUpliftWhileToForPatterns(RewritePatternSet &patterns);
 
+/// Populate patterns to rotate `scf.while` ops, constructing `do-while` loops
+/// from `while` loops.
+void populateSCFRotateWhileLoopPatterns(
+    RewritePatternSet &patterns, const SCFRotateWhileLoopPassOptions &options);
```

Comment:
This currently has a limitation: we cannot apply the pattern recursively on newly created ops in a pass when `forceCreateCheck`. I left it as is as I am not aware of what's the "canonical" approach here. I'm using `GreedyRewriteStrictness::ExistingOps` in the pass, but passes calling this may not wanna do that.

We could either document that here or add an anonymous attribute to output `scf.while` operations to avoid infinite recursion.

Reply:
I think annotating the operation might be a better option as it gives callers more flexibility. In my use-case, I don't care much as I won't be using `forceCreateCheck`, but others might be interested.

Reply:
Do you want to use pass itself or you plan to use `populateSCFRotateWhileLoopPatterns` to combine with other patterns?

If you only need pass, using direct loop like in test pass instead of greedy rewriter may be a better solution.

Reply:
I'm planning on using the patterns, that's why I went with this. We could also hide the problem by not allowing options in the patterns. The pass can remain as is. WDYT?

Reply:
Yeah, I feel not exposing this option on patterns is better solution, but you will need to keep old test pass to be able to test it.